core-bundle/src/Filesystem/FilesystemItemIterator.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+/**
+ * @implements \IteratorAggregate<int, FilesystemItem>
+ */
+class FilesystemItemIterator implements \IteratorAggregate
+{
+    /**
+     * @var iterable<FilesystemItem>
+     */
+    private iterable $listing;
+
+    /**
+     * @param iterable<FilesystemItem> $listing
+     */
+    public function __construct(iterable $listing)
+    {
+        $this->listing = $listing;
+    }
+
+    /**
+     * @param callable(FilesystemItem):bool $filter
+     */
+    public function filter(callable $filter): self
+    {
+        $listFiltered = static function (iterable $listing) use ($filter): \Generator {
+            foreach ($listing as $item) {
+                if ($filter($item)) {
+                    yield $item;
+                }
+            }
+        };
+
+        return new self($listFiltered($this->listing));
+    }
+
+    public function files(): self
+    {
+        return $this->filter(static fn (FilesystemItem $item) => $item->isFile());
+    }
+
+    public function directories(): self
+    {
+        return $this->filter(static fn (FilesystemItem $item) => !$item->isFile());
+    }
+
+    /**
+     * @return \Traversable<FilesystemItem>
+     */
+    public function getIterator(): \Traversable
+    {
+        return $this->listing instanceof \Traversable ?
+            $this->listing : new \ArrayIterator($this->listing);
```

Comment:
```suggestion
        return $this->filter(static fn (FilesystemItem $item) => true);
```
Type chek for `FilesystemItem` so that it is guaranteed that no other type is returned?

Reply:
This would end in an endless loop, though. Not sure if it's worth implementing a more sophisticated solution 'just' for the type check.

Reply:
Endless loop? 🤔

Reply:
https://3v4l.org/YNqUO#v8.1.2

Reply:
```suggestion
        foreach ($this->listing as $item){
            if (!$item instanceof FilesystemItem) {
                // Type error...
            }
            
            yield $item;
        }
```

Reply:
Nice! :sunglasses: 

Implemented in d5ca515a9b502952837f52232284511ff20849af